Feature: Implementations of Range-Separated Hybrid Functionals

1041176461 opened the pull request:
Range-separated hybrid (RSH) functionals are a class of hybrid functionals that partition the Fock exchange into short-range and long-range components:
![image](https://github.com/user-attachments/assets/31e14da8-38f9-494b-94e8-761e60d86f5b)
To address the divergence issue of the Coulomb potential at $q\to0$, two distinct RSH frameworks have been implemented:

1. Conventional Spencer-Alavi truncated Coulomb potential
   - More efficient in initialization,
   - but suffers from divergence when computing long-range corrected hybrid functionals (e.g., LC-PBE).

2. Massidda’s auxiliary-function-modified Ewald summation scheme
   -   Resolves the limitations of the Spencer-Alavi truncation method.

This computational framework serves as the foundation for the upcoming implementation of GW and RT-TDDFT-EXX functionalities.

1041176461 committed:
fix get_Gvec

1041176461 committed:
Fix: modify G calculations

1041176461 committed:
fix singular value

1041176461 committed:
fix cal_fq_type_1

1041176461 committed:
fix singularity

1041176461 committed:
add OpenMP reduction in lattice_sum

1041176461 committed:
add mpi for gauss_Vq

1041176461 committed:
MPI distribute

1041176461 committed:
fix compile bug 1

1041176461 committed:
fix compile bug 2

1041176461 committed:
accelerate lattice_sum

1041176461 committed:
add Rcut for list

1041176461 committed:
Merge branch 'deepmodeling:develop' into develop

1041176461 committed:
add mpi

1041176461 committed:
add default input for fq_type0

1041176461 committed:
fix fq_type_0

1041176461 committed:
fix ylm settings bug

1041176461 committed:
fix multipole bug

1041176461 committed:
add gauss Rcut

1041176461 committed:
Fix: delete custom Rmax for exx

1041176461 committed:
fix bug in initializing gauss

1041176461 committed:
modify gauss settings

1041176461 committed:
Refactor: abfs and gauss rcut settings

1041176461 committed:
Hard coded ewald_lambda

1041176461 committed:
Feature: add singluarity correction

1041176461 committed:
Fix: fq_type_1 bug

1041176461 committed:
Feature: Hard coded singular correction params

1041176461 committed:
format

1041176461 committed:
Fix: bug of MPI distribution

1041176461 committed:
modify default value for fq_type

1041176461 committed:
Fix: bug in Exx_LRI_interface

1041176461 committed:
set Nr to odd value

1041176461 committed:
Refactor: set chi and init parallel

1041176461 committed:
Fix: fix last refactor

1041176461 committed:
add Vws for cal_Vs_gauss

1041176461 committed:
Feature: add dVq gauss

1041176461 committed:
modify dVq dtype

1041176461 committed:
fix format error

1041176461 committed:
format

1041176461 committed:
delete const

1041176461 committed:
compile bug

1041176461 committed:
Fix: modify full_R to period_R

1041176461 committed:
Fix: check_cell match with RI_2D_Comm

1041176461 committed:
fix compile bug

1041176461 committed:
Refactor: gauss lattice sum

1041176461 committed:
Refactor: init ylm and use template

1041176461 committed:
Refactor: modify template data type

1041176461 committed:
Refactor: add template for cal_Vq_gauss

1041176461 committed:
Refactor: add template for cal_Vq_minus_gauss

1041176461 committed:
Refactor: add template for cal_Vq

1041176461 committed:
Refactor: add template for cal_Vs

1041176461 committed:
Fix: compile bug in gaussian_abfs

1041176461 committed:
Fix: compile bug

1041176461 committed:
Fix: convert bug

1041176461 committed:
Fix: add minus template

1041176461 committed:
Refactor: comm_map2_first for array<Tensor, 3>

1041176461 committed:
Fix: delete inline of get_Vq for external binding

1041176461 committed:
Fix: init bug

1041176461 committed:
Fix: skip gamma in lattice_sum

1041176461 committed:
Fix: Infinite loop in minus

1041176461 committed:
Fix: singular correction error

1041176461 committed:
Fix: init 2Lmax*1

1041176461 committed:
Fix: modify tools template

1041176461 committed:
Fix: save memory in init

1041176461 committed:
Fix: modify TITLE

1041176461 committed:
Refactor: omp init

1041176461 committed:
Fix: q to -q

1041176461 committed:
Fix: modify Rmax to Rmin for k convergency

1041176461 committed:
Fix: error tau index

1041176461 committed:
Fix: symmetry = 0 error

1041176461 committed:
Merge branch 'LC_Exx-new' of https://github.com/1041176461/abacus-develop into LC_Exx-new

1041176461 committed:
Fix: dVws flags error

1041176461 committed:
Fix: cal_dVs empty bug

1041176461 committed:
Fix: add waring

1041176461 committed:
Merge branch 'deepmodeling:develop' into develop

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

1041176461 committed:
Fix: merge error

1041176461 committed:
Fix: compile error

1041176461 committed:
Fix: cmake error

1041176461 committed:
Fix: cmake order

1041176461 committed:
Fix: ewald mpi error

1041176461 committed:
Fix: hard coded for fq_type 1

1041176461 committed:
Fix: add cam kernel for gauss part

1041176461 committed:
Refactor: omp for set_Vq_dVq_minus_gauss

1041176461 committed:
Refactor: optimize check_cell

1041176461 committed:
Refactor: delate move

1041176461 committed:
Refactor: check_cell

1041176461 committed:
Refactor: delete check_empty and add check_and_set_zero

1041176461 committed:
Fix: check_cell bug and delete set_zero

1041176461 committed:
Refactor: delete check_cell

1041176461 committed:
Fix: modify cal_latvec_range for other basis

1041176461 committed:
Refactor: match NAO and gauss rcut

1041176461 committed:
Merge branch 'LC_Exx-new' of https://github.com/1041176461/abacus-develop into LC_Exx-new

1041176461 committed:
Refactor: only keep cal_latvec_range fix

1041176461 committed:
Refactor: delete min Rcut

1041176461 committed:
Fix: still use min Rcut for keeping res under different lambda

1041176461 committed:
Fix: remove extra correction for sr kernel

1041176461 committed:
Fix: hf parameter error

1041176461 committed:
Refactor: move init_Gaunt out of Matrix_Orbs* init

1041176461 committed:
Refactor: add init_C  judging

1041176461 committed:
Refactor: set ewald init MGT

1041176461 committed:
Fix: ewald init false

1041176461 committed:
Fix: gauss dVq error

1041176461 committed:
Merge branch 'LC_Exx-new' of https://github.com/1041176461/abacus-develop into LC_Exx-new

1041176461 committed:
Fix: init_elem bug of std::array

1041176461 committed:
Refactor: comment dVs calculated by Ewald method

1041176461 committed:
Refactor: singular chi settings

1041176461 committed:
Merge branch 'deepmodeling:develop' into develop

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

1041176461 committed:
Refactor: add exx in tddft

1041176461 committed:
Fix: add exx_use_ewald param

1041176461 committed:
Fix: input param error

1041176461 committed:
Fix: param settings error

1041176461 committed:
Fix: bug in td-exx

1041176461 committed:
Merge branch 'deepmodeling:develop' into develop

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

ESROAMER committed:
Add files via upload

ESROAMER committed:
Update cal_edm_tddft.cpp

1041176461 committed:
Refactor: modify exx loop for relax/md

ESROAMER committed:
Update result.ref

ESROAMER committed:
Fix wrong parameters in integrate test

ESROAMER committed:
Update result.ref

ESROAMER committed:
Update result.ref

ESROAMER committed:
Update result.ref

ESROAMER committed:
Update result.ref

ESROAMER committed:
Update result.ref

ESROAMER committed:
Update result.ref

ESROAMER committed:
Update result.ref

ESROAMER committed:
Update result.ref

maki49 committed:
add exx when istep>0

1041176461 committed:
Merge pull request #8 from maki49/new_loop

Add exx when istep > 0

1041176461 committed:
Merge pull request #7 from 1041176461/new_loop

Refactor: modify exx loop for relax/md

1041176461 committed:
Merge branch 'new_loop' into LC_Exx-new

1041176461 committed:
Merge branch 'develop' into develop

1041176461 committed:
Update dos.py

1041176461 committed:
Update esolver_sdft_pw.cpp

1041176461 committed:
Update lcao_before_scf.cpp

1041176461 committed:
Update Exx_LRI_interface.h

1041176461 committed:
Update Exx_LRI_interface.hpp

1041176461 committed:
Fix: compile error

1041176461 committed:
Fix: compile error

1041176461 committed:
Fix: change HSE relax/md result.ref for new framework

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

1041176461 committed:
Merge branch 'pr/1041176461/9' into develop

1041176461 committed:
Fix: compile error

1041176461 committed:
compatible with exx_iter_finish

ESROAMER committed:
Merge branch 'deepmodeling:develop' into develop

ESROAMER committed:
Add files via upload

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

ESROAMER committed:
Update esolver_ks_lcao_tddft.h

ESROAMER committed:
Update esolver_ks_lcao_tddft.cpp

1041176461 committed:
Merge pull request #10 from ESROAMER/no_force

merge from ESROAMER:no_force

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

ESROAMER committed:
hybrid

ESROAMER committed:
Merge branch 'hytd' into LC_Exx-new

1041176461 committed:
Refactor: add input check for TD-EXX

1041176461 committed:
Merge branch 'LC_Exx-new' of https://github.com/1041176461/abacus-develop into LC_Exx-new

1041176461 committed:
Revert hytd commit

1041176461 committed:
Fix: input error

1041176461 committed:
Fix: input error

1041176461 committed:
Fix: input error

1041176461 committed:
add istep check for cal_exx_ions

1041176461 committed:
delete exx_ewald_qdiv

1041176461 committed:
Fix: pbe0 bug and delete cam_alpha and cam_beta param

1041176461 committed:
Merge branch 'deepmodeling-develop' into develop

1041176461 committed:
Merge branch 'develop' of https://github.com/1041176461/abacus-develop into develop

1041176461 committed:
Feature: add a new Rcut for HF

1041176461 committed:
Fix: add param Rcut_type

1041176461 committed:
Merge branch 'deepmodeling:develop' into develop

1041176461 committed:
Fix: support negative value in parse_expression

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

1041176461 committed:
Fix: param settings error

1041176461 committed:
Fix: add Rcut for set_Vs_dVs

1041176461 committed:
Fix: mixing pulay

1041176461 committed:
Fix: large system rcut error

1041176461 committed:
Merge branch 'deepmodeling:develop' into develop

1041176461 committed:
Fix: parse_expression for scientific notation

1041176461 committed:
Update input_conv.h

1041176461 committed:
Merge branch 'deepmodeling:develop' into develop

1041176461 committed:
fix kvec_c_full for mpi

1041176461 committed:
Merge branch 'deepmodeling:develop' into develop

1041176461 committed:
merge develop

1041176461 committed:
fix compile error

1041176461 committed:
fix ucell error

1041176461 committed:
fix init_ions

1041176461 committed:
fix cell

1041176461 committed:
fix k for symmetry=0

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

1041176461 commented:
The test suite and documentation are currently being prepared.

1041176461 committed:
namespace Singular_Value

1041176461 committed:
Merge branch 'LC_Exx-new' of https://github.com/1041176461/abacus-develop into LC_Exx-new

mohanchen commented on source/Makefile.Objects:
```diff
@@ -640,6 +640,8 @@ OBJS_MODULE_RI=conv_coulomb_pot_k.o\
       symmetry_rotation.o\
       symmetry_irreducible_sector.o\
+      auxiliary_func.o\
+      ewald_Vq_tools.o\
```
I would suggest using 'vq' instead of 'Vq'

mohanchen commented on source/module_cell/klist.h:
```diff
@@ -13,6 +13,7 @@ class K_Vectors
 public:
     std::vector<ModuleBase::Vector3<double>> kvec_c; /// Cartesian coordinates of k points
     std::vector<ModuleBase::Vector3<double>> kvec_d; /// Direct coordinates of k points
+    std::vector<ModuleBase::Vector3<double>> kvec_c_full; // Cartesian coordinates of full k mesh match with nkstot_full
```
what' the definition of 'full' in this case?

1041176461 replied:
The current implementation of the Ewald summation for calculating the EXX Vq term does not support symmetry operations, requiring the use of the full k-point grid without symmetry reduction. This "full" keyword can be removed once symmetry support is implemented in future updates.

mohanchen replied:
I would suggest you open a local array to store all kvec_c_full data in the code where you need it, do not add new iterms in klist.h, because your code will introduce uncessary k-point arryas for other calculations.

1041176461 replied:
@maki49 Could you give me some advice? The currently existing k-points have been symmetry-reduced. How should I store kvec_c_full as a local variable before the reduction? Or do I need to generate another set of k-points and then communicate them to different processes?

maki49 replied:
@1041176461 @mohanchen `kvec_c_full` has to be a member because it is used outside of class `K_Vectors` (in ewald_Vq.hpp at line 111). I think it's resonable to store the unreduced k-points as a member of `K_Vectors` under the current design. (A better design will make it possible to have two different objects :`K_Vectors kv; K_Vectors kv_full;`.)

1041176461 committed:
fix text_xc1

mohanchen commented on source/module_hamilt_general/module_xc/xc_functional_libxc.cpp:
```diff
@@ -153,16 +153,70 @@ std::vector<xc_func_type> XC_Functional_Libxc::init_func(const std::vector<int>
 		{
 			add_func( XC_GGA_X_ITYH );
-			double parameter_omega[1] = {PARAM.inp.exx_hse_omega}; // GlobalC::exx_info.info_global.hse_omega
+			double parameter_omega[1] = {GlobalC::exx_info.info_global.hse_omega}; // GlobalC::exx_info.info_global.hse_omega
```
Please don't use GlobalC

1041176461 replied:
This follows the original coding conventions. Replacing GlobalC would entail substantial refactoring of the existing framework in [input_conv.cpp ](https://github.com/deepmodeling/abacus-develop/blob/develop/source/module_io/input_conv.cpp), which is beyond the intended scope of this PR

mohanchen replied:
You can use PARAM instead of GlobalC

1041176461 replied:
Deleted

mohanchen commented on source/module_ri/Faddeeva.cc:
the file is too long

1041176461 replied:
Deleted

mohanchen commented on source/module_ri/Faddeeva.hh:
what's the code used for?

1041176461 replied:
I need to use the imaginary error function erfi(x) that supports complex input x, but the standard std::erf function only accepts input x of type double and cannot be used to construct the erfi function. Currently, the only numerical implementation I found is the Faddeeva package (http://jdj.mit.edu/faddeeva/), and the erfi function in Scipy is a wrapper around this package.

1041176461 replied:
The protocol used in this code might differ from that of ABACUS. I've reviewed the implementation, and it appears to be quite lengthy. Do you think there are better alternatives available, or should we consider making some modifications before integrating it into ABACUS?

1041176461 replied:
Add a new class for calculating erf(x)、erfi(x)、erfc(x)  that supports complex input x.

mohanchen commented on tools/rt-tddft-tools/examples/ground-state-projection-Si/Excited_band_for_ik1.jpeg:
Currently we don't need jpeg file in the software

mohanchen commented on tools/rt-tddft-tools/examples/Absorpation-N2/out_dipole.png:
Currently we don't need jpeg file in the software

mohanchen commented on tools/rt-tddft-tools/examples/Absorpation-N2/out_abs.png:
Currently we don't need jpeg file in the software

1041176461 committed:
add doc

1041176461 committed:
delete Excited_band_for_ik1.jpeg

1041176461 committed:
Delete out_abs.png

1041176461 committed:
Delete out_dipole.png

1041176461 committed:
Delete tools/rt-tddft-tools/examples/Absorpation-N2/out_spectra.dat

1041176461 committed:
Delete tools/rt-tddft-tools/examples/Absorpation-N2/SPIN1_DIPOLE

1041176461 committed:
Delete tools/rt-tddft-tools/examples/Absorpation-N2/efield_0.dat

1041176461 committed:
Delete tools/rt-tddft-tools/examples/Absorpation-N2/efield_1.dat

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

1041176461 committed:
modify ewald_Vq to ewald_vq

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

1041176461 committed:
delete GlobalC

1041176461 committed:
Merge branch 'LC_Exx-new' of https://github.com/1041176461/abacus-develop into LC_Exx-new

1041176461 committed:
fix compile

1041176461 committed:
fix test

1041176461 committed:
fix test_xc

1041176461 committed:
fix test

1041176461 committed:
add complex ErrorFunc

1041176461 committed:
delete Faddeeva

1041176461 committed:
delete

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

1041176461 committed:
add tests for two implementations

1041176461 committed:
delete unused functions

1041176461 commented:
> The test suite and documentation are currently being prepared.

Added

1041176461 committed:
add tests

1041176461 committed:
fix tests

1041176461 committed:
set nosym

1041176461 committed:
fix tests

1041176461 committed:
fix tests

1041176461 committed:
fix tests

1041176461 committed:
Merge branch 'LC_Exx-new' of https://github.com/1041176461/abacus-develop into LC_Exx-new

All conflicts fixed but you are still merging.

1041176461 committed:
fix txt

1041176461 committed:
fix dir name

1041176461 committed:
fix doc

1041176461 committed:
fix tests

1041176461 committed:
no limits for hse_omega and hybrid_beta

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

dyzheng commented on source/module_ri/ewald_Vq.h:
```diff
@@ -0,0 +1,219 @@
+#include <mpi.h>
+
+template <typename Tdata>
+class Ewald_Vq
```
There are no notes to introduce this class, I can't know the algorithm for this, can you add some brief introduction or any reference?

dyzheng commented on source/module_ri/Matrix_Orbs21.cpp:
```diff
@@ -9,15 +9,17 @@
-                         const double kmesh_times, 
-                         const double rmesh_times)
+void Matrix_Orbs21::init(
+	const int mode, 
```
which format tools did you used? it looks invalid for Indentation.

dyzheng commented on source/module_ri/Matrix_Orbs21.cpp:
```diff
@@ -45,19 +47,22 @@ void Matrix_Orbs21::init(const int mode,
-    this->MGT.init_Gaunt(2 * Lmax + 1);
+    // //=========================================
+    // // (3) make Gaunt coefficients table
+    // //=========================================
```
It seems to be an accidental operation.

dyzheng commented on source/module_ri/RPA_LRI.h:
```diff
@@ -79,7 +80,7 @@ template <typename T, typename Tdata> class RPA_LRI
     // Tdata post_process_Erpa( const Tdata &Erpa_in ) const;
 };
-Exx_LRI<double> exx_lri_rpa(GlobalC::exx_info.info_ri);
+Exx_LRI<double> exx_lri_rpa(GlobalC::exx_info.info_ri, GlobalC::exx_info.info_ewald);
```
Please add namespace for this global class.

dyzheng replied:
The class is also need to defined in namespace.

1041176461 replied:
The RPA part is not my work; I only modified the initialization code. Further modifications may need to be done by @AroundPeking in his GW implementations

AroundPeking replied:
Add namespace for `Exx_LRI` and `RPA_LRI`?
Can this pull request be merged without adding a namespace? The source before it was merged also did not have one.
The namespace can be added later.

dyzheng commented on source/module_ri/conv_coulomb_pot_k.h:
```diff
@@ -1,17 +1,22 @@
+		Erf,    //  	"hse_omega", "hf_Rcut"
+        Cam,    //  	"hse_omega", "hybrid_alpha", "hybrid_beta", "hf_Rcut"
+		Ccp_Cam // "hse_omega", "hybrid_alpha", "hybrid_beta"
+    };
```
Please add annotations for this class and functions.

dyzheng commented on source/module_ri/gaussian_abfs.h:
```diff
@@ -0,0 +1,117 @@
+#include <map>
+#include <vector>
+
+class Gaussian_Abfs
```
Please add annotations for this class and functions.

dyzheng commented:
Congratulations on your outstanding work!
From the perspective of long-term code maintenance, I suggest you add sufficient comments to enable other developers to understand the functionality of this part of the code. Unless you promise to respond and maintain all issues related to the code.

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

1041176461 committed:
delete GlobalC in libxc

1041176461 committed:
Merge branch 'LC_Exx-new' of https://github.com/1041176461/abacus-develop into LC_Exx-new

1041176461 committed:
add notes for Ewald_Vq

1041176461 committed:
format

1041176461 committed:
add notes for Ccp_Type

1041176461 committed:
add notes

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

mohanchen commented:
Hi, you may need to resolve the conflicts, thanks for your efforts! @1041176461

1041176461 committed:
Merge branch 'deepmodeling:develop' into develop

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

1041176461 committed:
delete info_ewald

1041176461 committed:
fix merge error

1041176461 committed:
move hybrid_alpha into ccp settings

1041176461 committed:
fix compile error

1041176461 committed:
fix compile error

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

1041176461 commented:
I moved mixing parameters into "conv_coulomb_pot_k.h" and changed short-range hybrid functionals mixing settings  to match with range-separated hybrid functionals. @maki49 @AroundPeking @PeizeLin @Flying-dragon-boxing @JGHan7 Please check whether the current modification will affect your code (see commit https://github.com/deepmodeling/abacus-develop/pull/6220/commits/9e9c5e8f28630c4149531d4f0899df121403b7c8). Thanks.

1041176461 committed:
fix tests

1041176461 committed:
Merge branch 'LC_Exx-new' of https://github.com/1041176461/abacus-develop into LC_Exx-new

1041176461 committed:
fix ewald param

1041176461 committed:
delete exx_hybrid_beta in read_input

1041176461 committed:
Merge branch 'LC_Exx-new' of https://github.com/1041176461/abacus-develop into LC_Exx-new

1041176461 committed:
modify exx_v_threshold in tests for new settings

1041176461 committed:
Merge branch 'develop' into LC_Exx-new

mohanchen commented:
Reopen it if needed
